File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendFeedbackRepository.kt
package com.humblesolutions.indsphinx.repository

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.humblesolutions.indsphinx.model.Feedback
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import java.util.Date

class BackendFeedbackRepository : FeedbackRepository {
    private val db = FirebaseFirestore.getInstance()

    override suspend fun submitFeedback(feedback: Feedback): String {
        val data = hashMapOf(
            "OccupantId" to feedback.occupantId,
            "OccupantName" to feedback.occupantName,
            "title" to feedback.title,
            "description" to feedback.description,
            "date" to Timestamp(Date())
        )
        val doc = db.collection("Feedback").add(data).await()
        return doc.id
    }

    override fun observeByOccupant(occupantId: String): Flow<List<Feedback>> = callbackFlow {
        val registration = db.collection("Feedback")
            .whereEqualTo("OccupantId", occupantId)
            .orderBy("date", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null || snapshot == null) { trySend(emptyList()); return@addSnapshotListener }
                val feedbacks = snapshot.documents.mapNotNull { doc ->
                    val title = doc.getString("title") ?: return@mapNotNull null
                    Feedback(
                        id = doc.id,
                        occupantId = doc.getString("OccupantId") ?: "",
                        occupantName = doc.getString("OccupantName") ?: "",
                        title = title,
                        description = doc.getString("description") ?: "",
                        date = (doc.get("date") as? Timestamp)?.toDate()?.time ?: 0L
                    )
                }
                trySend(feedbacks)
            }
        awaitClose { registration.remove() }
    }
}

File: iosApp/iosApp/BackendFeedbackRepository.swift
import Foundation
import FirebaseFirestore

struct FeedbackItem: Identifiable, Hashable {
    let id: String
    let occupantId: String
    let occupantName: String
    let title: String
    let description: String
    let date: Date
}

class BackendFeedbackRepository {
    private let db = Firestore.firestore()

    func submitFeedback(
        occupantId: String,
        occupantName: String,
        title: String,
        description: String
    ) async throws -> String {
        let data: [String: Any] = [
            "OccupantId": occupantId,
            "OccupantName": occupantName,
            "title": title,
            "description": description,
            "date": Timestamp(date: Date())
        ]
        let ref = try await db.collection("Feedback").addDocument(data: data)
        return ref.documentID
    }

    func observeByOccupant(occupantId: String, onChange: @escaping ([FeedbackItem]) -> Void) -> ListenerRegistration {
        return db.collection("Feedback")
            .whereField("OccupantId", isEqualTo: occupantId)
            .order(by: "date", descending: true)
            .addSnapshotListener { snapshot, _ in
                guard let snapshot = snapshot else { onChange([]); return }
                let feedbacks: [FeedbackItem] = snapshot.documents.compactMap { doc in
                    let data = doc.data()
                    guard let title = data["title"] as? String else { return nil }
                    return FeedbackItem(
                        id: doc.documentID,
                        occupantId: data["OccupantId"] as? String ?? "",
                        occupantName: data["OccupantName"] as? String ?? "",
                        title: title,
                        description: data["description"] as? String ?? "",
                        date: (data["date"] as? Timestamp)?.dateValue() ?? Date()
                    )
                }
                onChange(feedbacks)
            }
    }
}
